src/articulation.js

```diff
@@ -12,12 +12,95 @@
 import { Vex } from './vex';
 import { Flow } from './tables';
 import { Modifier } from './modifier';
-import { StaveNote } from './stavenote';
 import { Glyph } from './glyph';
+import { Stem } from './stem';
 
 // To enable logging for this class. Set `Vex.Flow.Articulation.DEBUG` to `true`.
 function L(...args) { if (Articulation.DEBUG) Vex.L('Vex.Flow.Articulation', args); }
 
+const { ABOVE, BELOW } = Modifier.Position;
+
+const mathToIncrement =
+  (mathFn, increment, value) =>
+    mathFn(value / increment) * increment;
+
+const isWithinStaff = (line) => (1 <= line && line <= 5);
+
+const snapLineToStaff = (canSitBetweenLines, line, position, offsetDirection) => {
+  const adjustmentFunction = position === ABOVE ? Math.ceil : Math.floor;
+
+  // Snap To staff line or staff space
+  let snappedLine = mathToIncrement(adjustmentFunction, 0.5, line);
+  const onStaffLine = Math.abs(snappedLine % 1) === 0;
+
+  const shouldSnapToStaffSpace =
+    canSitBetweenLines && isWithinStaff(snappedLine) && onStaffLine;
+
+  // If within staff, snap to staff space
+  if (shouldSnapToStaffSpace) {
+    snappedLine += 0.5 * -offsetDirection;
+  }
+
+  return snappedLine;
+};
+
+const getTopY = (note, textLine) => {
+  const stave = note.getStave();
+  const stemDirection = note.getStemDirection();
+  const { topY: stemTipY, baseY: stemBaseY } = note.getStem().getExtents();
+
+  if (note.getCategory() === 'stavenotes') {
+    if (stemDirection === Stem.UP) {
+      return stemTipY;
+    } else {
+      return stemBaseY;
+    }
+  } else if (note.getCategory() === 'tabnotes') {
+    if (note.hasStem()) {
+      if (stemDirection === Stem.UP) {
+        return stemTipY;
+      } else {
+        return stave.getYForTopText(textLine - 1);
+      }
+    } else {
+      return stave.getYForTopText(textLine - 1);
+    }
+  } else {
+    throw new Vex.RERR(
+      'UnknownCategory', 'Only can get the top and bottom ys of stavenotes and tabnotes'
+    );
+  }
+};
+
+const getBottomY = (note, textLine) => {
+  const stave = note.getStave();
+  const stemDirection = note.getStemDirection();
+  const { topY: stemTipY, baseY: stemBaseY } = note.getStem().getExtents();
+
+  if (note.getCategory() === 'stavenotes') {
+    debugger;
+    if (stemDirection === Stem.UP) {
+      return stemBaseY;
+    } else {
+      return stemTipY;
+    }
+  } else if (note.getCategory() === 'tabnotes') {
+    if (note.hasStem()) {
+      if (stemDirection === Stem.UP) {
+        return stave.getYForBottomText(textLine - 2);
+      } else {
+        return stemTipY;
+      }
+    } else {
+      return stave.getYForBottomText(textLine - 2);
```

Comment:
I guess this is because of the `bottom_text_position` default being `numLines + 1`?